packages/snaps-execution-environments/src/common/endowments/abortController.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * Creates wrapped AbortController endowment hardened by SES.
+ *
+ * @returns An object with the attenuated `AbortController` function.
+ */
+const createAbortController = () => {
+  /**
+   * Wrapped and attenuated AbortController.
+   */
+  const _AbortController = class implements AbortController {
+    #abortController: AbortController;
+
+    readonly signal: AbortSignal;
+
+    #signalHandler: ProxyHandler<object>;
+
+    constructor() {
+      this.#signalHandler = {
+        get(target: AbortSignal, prop: string) {
+          if (prop === 'aborted') {
+            return target.aborted;
+          } else if (prop === 'onabort') {
+            return target.onabort;
+          } else if (prop === 'reason') {
+            return target.reason;
+          } else if (prop === 'throwIfAborted') {
+            return () => target.throwIfAborted();
+          }
+          return undefined;
+        },
+      };
+      this.#abortController = new AbortController();
+      // @ts-expect-error This needs to be attenuated object rather than real
+      this.signal = new Proxy(
```

Comment:
To make it easier to reason about the signal object should be an instance of our dummy _AbortSignal class.
The handling of the signal itself needs a factory that binds an instance to an actual AbortSignal instance that is never accessible inside a snap but our _AbortSignal instance can be exchanged for the related AbortSignal instance outside of the snap.

That is done via a WeakMap where _AbortSignal instance is the key and AbortSignal instance is the value.
ALongside the factory, you also need a function that does the exchange. they both need to be kept private to our implementation only and share a scope in which the WeakMap lives.

Reply:
Thanks for the information provided.
But, as we agreed on slack this will be done in the next ticket. Because of the complexity for this case it is the best to keep this work separate.
The work will be followed in this ticket: https://github.com/MetaMask/snaps-monorepo/issues/1132